python/ray/data/_internal/execution/operators/base_physical_operator.py

```diff
@@ -94,23 +94,34 @@ def _add_input_inner(self, refs: RefBundle, input_index: int) -> None:
         assert not self.completed()
         assert input_index == 0, input_index
         self._input_buffer.append(refs)
+        self._metrics.on_input_queued(refs)
 
     def all_inputs_done(self) -> None:
         ctx = TaskContext(
             task_idx=self._next_task_index,
             sub_progress_bar_dict=self._sub_progress_bar_dict,
             target_max_block_size=self.actual_target_max_block_size,
         )
-        self._output_buffer, self._stats = self._bulk_fn(self._input_buffer, ctx)
+
+        outputs, self._stats = self._bulk_fn(self._input_buffer, ctx)
+        for output in outputs:
+            self._output_buffer.append(output)
+            self._metrics.on_output_queued(output)
+
         self._next_task_index += 1
-        self._input_buffer.clear()
+
+        while self._input_buffer:
+            input = self._input_buffer.pop()
+            self._metrics.on_input_dequeued(input)
```

Comment:
This doesn't account for the intermediate blocks. 
But it probably doesn't matter because of the other comment I left.

Reply:
Fixed